### DISCUSSION TITLE
How doe this tasks extension know what is a test?

### DISCUSSION THREAD

**Discussion by lsantos10** [microsoft/PR-Metrics#120] [Q&A] [answered]
When adding this to my PRs it provides as table with number of rows and number of rows tested.
I see that we have an option for setting what are the code file patterns, but how does it know what is a test?

**Comment by muiriswoulfe** [comment, marked as answer]
Hi @lsantos10.

The logic for determining whether a file is a test is as follows:

> Files with `test` in the file or directory name (irrespective of case) are considered test files. All other files are considered product code files.

PR Metrics is designed to give a quick overview of the metrics for all coding languages, so only the file name and lines changed are checked by the extension (via the Git command `git diff --numstat origin/<target>...pull/<pull_request_id>/merge`). It avoids analysing the content of the file as this would require a slower language-specific analysis and there are other tools available that can perform this task. But it performs the quick check I outlined above, so that you can get an estimate of the balance of product and test code. The key is to consider this a quick datapoint for driving and reinforcing positive PR behaviours, and to use other tooling if you need something more authoritative such as code coverage.

This regular expression is not currently available as a setting for the purposes of simplifying how the extension is used (especially in the classic, UX-based pipelines). It was found that the current design addresses the vast majority of test code without the complexity of an extra setting.

As I mentioned, we believe that the current pattern addresses the vast majority of tests, but if there is some other regular expression you believe should be added to cover more test files, let me know and I'll see what we can do. If you really believe there is a need for a setting, also let me know and I can also follow up on this. In the latter case, I'd really like to understand your test setup better to understand the need for this though, so that we can avoid the extra complexity if there is an another solution available. But ultimately, we do want to deliver a solution that works for everyone within the constraints of keeping it fast and simple, so I do want to hear back if you have ideas.

More information on the test file determination logic can be found at <https://github.com/microsoft/OMEX-Azure-DevOps-Extensions/tree/main/PipelinesTasks/PRMetrics#implementation>.

**Reply by lsantos10** [reply]
Well, for angular application with karma tests (which is pretty common), the product code lives on *.ts file and the tests live on same directory and same name, just changing the extension to *.spec.ts. 
For this case, we are unable to have the extension providing the expected result

If we had the option to define the pattern for tests as well this would be fixed.

**Reply by muiriswoulfe** [reply]
Thanks for clarifying. I think including `*.spec.*` as a test match string makes sense. I'll make the update for this soon and let you know when it's available.

**Reply by muiriswoulfe** [reply]
I've now made the changes at #121. I'll update this thread after the change gets published out to the Azure DevOps Marketplace.

**Reply by muiriswoulfe** [reply]
It's now in the Marketplace as Release 1.2.6. You should be automatically upgraded to it as it's only a minor revision.

Let me know if you encounter any issues.
